src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs

```diff
@@ -0,0 +1,360 @@
+            if (state != null && 
+                state.RequestHandle != null &&
+                !state.RequestHandle.IsInvalid &&
+                state.RequestHandle.DangerousGetHandle() == handle)
+            {
+                invokeCallback = true;
+            }
+
+            WinHttpTraceHelper.TraceStatusCallback("WinHttpCallback", handle, invokeCallback, internetStatus);
+
+            if (invokeCallback)
+            {
+                RequestCallback(handle, state, internetStatus, statusInformation, statusInformationLength);
+            }
+        }
+        
+        private static void RequestCallback(
+            IntPtr handle,
+            WinHttpRequestState state,
+            uint internetStatus,
+            IntPtr statusInformation,
+            uint statusInformationLength)
+        {
+            try
+            {
+                switch (internetStatus)
+                {
+                    case Interop.WinHttp.WINHTTP_CALLBACK_STATUS_SENDREQUEST_COMPLETE:
+                        OnRequestSendRequestComplete(state);
+                        return;
+                        
+                    case Interop.WinHttp.WINHTTP_CALLBACK_STATUS_READ_COMPLETE:
+                        OnRequestReadComplete(state, statusInformationLength);
+                        return;
+                        
+                    case Interop.WinHttp.WINHTTP_CALLBACK_STATUS_WRITE_COMPLETE:
+                        OnRequestWriteComplete(state);
+                        return;
+                        
+                    case Interop.WinHttp.WINHTTP_CALLBACK_STATUS_HEADERS_AVAILABLE:
+                        OnRequestReceiveResponseHeadersComplete(state);
+                        return;
+
+                    case Interop.WinHttp.WINHTTP_CALLBACK_STATUS_REDIRECT:
+                        string redirectUriString = Marshal.PtrToStringUni(statusInformation);
+                        var redirectUri = new Uri(redirectUriString);
+                        OnRequestRedirect(state, redirectUri);
+                        return;
+
+                    case Interop.WinHttp.WINHTTP_CALLBACK_STATUS_SENDING_REQUEST:
+                        OnRequestSendingRequest(state);
+                        return;
+
+                    case Interop.WinHttp.WINHTTP_CALLBACK_STATUS_REQUEST_ERROR:
+                        Debug.Assert(
+                            statusInformationLength == Marshal.SizeOf<Interop.WinHttp.WINHTTP_ASYNC_RESULT>(),
+                            "RequestCallback: statusInformationLength=" + statusInformationLength +
+                            " must be sizeof(WINHTTP_ASYNC_RESULT)=" + Marshal.SizeOf<Interop.WinHttp.WINHTTP_ASYNC_RESULT>());
+
+                        var asyncResult = Marshal.PtrToStructure<Interop.WinHttp.WINHTTP_ASYNC_RESULT>(statusInformation);
+                        OnRequestError(state, asyncResult);
+                        return;
+
+                    default:
```

Comment:
We don't need to explicitly handle the `WINHTTP_CALLBACK_STATUS_SECURE_FAILURE` error. Although, it does have more details about the SSL/TLS failure which is good for logging. In the end, we'll get SECURE_FAILURE error back from WinHTTP.

The current WinHttpHandler is not reacting to `WINHTTP_CALLBACK_STATUS_HANDLE_CLOSING` callbacks.  The SafeHandle is already closed. WinHttpHandler doesn't have the same SafeHandle pattern as WebSockets.Client.